### ISSUE TITLE
Can't get Homebridge Thermostat status.

### ISSUE THREAD

**Issue by meluvalli** [oznu/homebridge-gsh#24]
I have HomeBridge installed and running with a Z-Wave thermostat.  I can control the thermostat fine through HomeKit, however, google doesn't work with it.  

Homebridge shows: "[Google Smart Home] Received QUERY intent", however the Google Home app shows it's Zero degrees and no status.

* **Node.js Version**: 13.2.0
* **NPM Version**: 6.13.2
* **Homebridge Version**: 2.4.0
* **Homebridge GSH Plugin Version**: 1.1.0
* **Homebridge Config UI X Plugin Version**: 4.6.5
* **Operating System**: Windows

**Comment by oznu** [comment]
Hi @meluvalli.

Can you enable debug mode in the plugin settings and then provide the logs?

After you have enabled debug mode, and restarted Homebridge, I'll need to see the SYNC event (pull down in the Google Home app to trigger this) and the QUERY request/response.

**Comment by oznu** [comment]
Ok I have checked the logs and it might be caused by Google Home not liking the current target temperature not being reported in .5 increments.   

```json
{ 
  "online":true, 
  "thermostatMode":"cool", 
  "thermostatTemperatureSetpoint":23.88, 
  "thermostatTemperatureAmbient":22.7
}
```

I imagine this strange value is caused by some F to C conversion (HomeKit uses celsius in the backend even if the front end is showing the temperature in fahrenheit.

Before I go making any code changes, do you have the ability to set your thermostat to a round number in celsius? And once that's done can you try QUERY the status again?

**Comment by meluvalli** [comment]
Hello.  I unfortunately have no way of changing that (that I know of anyways).  I have a Honeywell Z-Wave thermostat.  I am not sure how I would change that I should say :(.

I tried setting in HomeSeer to display in Celsius, but it didn't change it.  Not sure if maybe I have to delete and re-add it?

**Comment by meluvalli** [comment]
Ok.  I put my thermostat in Celsius mode and manually changed it that way.  Now the Thermostat in Google shows Indoor "-4.5" and it is showing it's set for "Cool" and it's set for "1".  

I have it set for cool, and it's currently 23.5 degrees Celsius in my home and I have it set to 34.0.

So, some progress as at least it's now showing something in Google, but still doesn't seem right.

**Comment by oznu** [comment]
Great, that tells me what I need to know I think; you can swap everything back.

This will require a plugin update to fix that I'll try and get to in the next couple days.

**Comment by meluvalli** [comment]
`  {
    id: 'XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX',
    type: 'action.devices.types.THERMOSTAT',
    traits: [ 'action.devices.traits.TemperatureSetting' ],
    attributes: {
      availableThermostatModes: 'off,heat,cool,auto',
      thermostatTemperatureUnit: 'C'
    },
    name: { defaultNames: [Array], name: 'Thermostat', nicknames: [] },
    willReportState: true,
    deviceInfo: { manufacturer: 'HomeSeer', model: 'Z-Wave Sensor Multilevel' },
    customData: {
      aid: 76,
      iid: 8,
      instanceUsername: 'XX:XX:XX:XX:XX:XX',
      instanceIpAddress: '169.254.219.70',
      instancePort: 51826
    }`

And from Resync:

`    "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX": {
      "online": true,
      "thermostatMode": "cool",
      "thermostatTemperatureSetpoint": 25.55,
      "thermostatTemperatureAmbient": -4.7
    }`

Hope this helps.  Not sure if the ID's are crucial to keep private, so I X'ed them out.

**Comment by meluvalli** [comment]
Ok, I know this isn't the best way of doing it, but I temporarily fixed it on my own system by forcing it to use Fahrenheit in the thermostat.js code.

Changed the line:
`thermostatTemperatureUnit: service.characteristics.find(x => x.type === hap_types_1.Characteristic.TemperatureDisplayUnits).value ? 'F' : 'C',`

to:
`thermostatTemperatureUnit: 'F',`

Is there a way for me to specify Fahrenheit in the configuration file?  Maybe I'm missing something?

After making that change, it's working 100%, but I know this is a HACK fix, so would prefer something more solid!

**Comment by meluvalli** [comment]
Also, on a side note, Not sure if this is a plug-in issue or Google Issue, but as you could see from my RESYNC, thermostat modes are off,heat,cool,auto.   My thermostat supports auto, but in Google, Auto isn't an option.  Only have off, heat, cool.

**Comment by oznu** [comment]
Interesting, this should be set from HomeKit, is the thermostat temperature units set to F in the iOS Home app? This is different to the units shown in the Home app (strangely enough). 

Go to Home -> Open up the thermostat properties -> Hardware Display -> Degrees C / F?

**Comment by oznu** [comment]
![54B21B32-6849-4927-BAF9-F5D915A7409D](https://user-images.githubusercontent.com/3979615/70956407-07862000-20c8-11ea-848d-d49ad847d331.jpeg)

**Comment by meluvalli** [comment]
Yes. I confirmed I have in HomeKit set to Fahrenheit for my thermostat.  

![Screenshot 2019-12-16 at 10 57 54 PM](https://user-images.githubusercontent.com/8729476/70966105-a0308780-2057-11ea-85cc-f2763f994e00.png)

**Comment by oznu** [comment]
Ok, that's where I read the F or C value from that you've manually set. I'll double check this is working as intended.

**Comment by oznu** [comment]
In regards to "AUTO" mode; it appears the visual control on the Google Home app does not support this. You can use voice to say "Set thermostat to auto" and it will work as intended.

**Comment by oznu** [comment]
Hi @meluvalli,

I can't work out why the `homebridge-homeseer` plugin is reporting the incorrect degree unit type. I've published a test version of the plugin that will log the value of the `TemperatureDisplayUnits` to your Homebridge logs. Could you give this test version a go and send me the results?

You will see something like this in the logs:

```
GSH - THERMOSTAT DISPLAY UNITS: {"iid":14,"type":"00000036-0000-1000-8000-0026BB765291","perms":["pr","pw","ev"],"format":"uint8","value":0,"description":"Temperature Display Units","valid-values":[0,1],"maxValue":1,"minValue":0}
```

You can install the test version using the command line and the `test` tag:

```bash
# windows 10
npm install -g homebridge-gsh@test

# most linux installs
sudo npm install -g homebridge-gsh@test

# docker
npm install homebridge-gsh@test
```

This `test` version also includes support for what Google Home calls "Heat + Cool Mode", which works like "Auto" does in HomeKit - you can set a temperature range if your thermostat supports it.

**Comment by meluvalli** [comment]
Ok.  Test installed.  Here we are:

<pre>GSH - THERMOSTAT DISPLAY UNITS: {"iid":13,"type":"XXXXXXXXXXXXXXXXXXXXXXXXXXXX","perms":["pr","pw","ev"],"format":"uint8","value":0,"description":"Temperature Display Units","valid-values":[0,1],"maxValue":1,"minValue":0}</pre>

Wow!  That took me <code>FOREVER</code> just to figure out how you put it inside a scroll able window!  Learn something new everyday!  ROFL!  Don't post much on here, so forgive me :)  I feel so old :(

**Comment by meluvalli** [comment]
Ok....  Found more information... I may know the issue here and it may not have anything to do with your plug in.   I went back in and looked at my Unit for HomeKit and it's now showing Celsius.  I changed it.  And in my log window I get "*Warning * -  Changing Hardware Display Units Does Nothing!".   Closed homebridge, re-opened it and it goes back to Celsius automatically.  It doesn't hold the setting.

I am not sure what would cause it to default to Celsius.  I'm running Homeseer and it's set for Fahrenheit...  So i'm confused :(

**Comment by oznu** [comment]
That's what I expected to happen after looking at the code for the plugin.

I'll try and sort something out tomorrow.

**Comment by meluvalli** [comment]
Ok.  Thank you so much for your help!

**Comment by meluvalli** [comment]
Update:
I found out that HomeBridge-HomeSeer-2018 is updating to a newer version...  In the process, they have/are fixing the Fahrenheit issue.  I have tested this and it now makes this plugin work correctly (kind of)...

Google now sees it correctly in the Google Home app.  However, there is still an issue.  If you set the thermostat to lets say 78F through google, it actually sets it to 79F.   The reason being is because it's rounding I assume.  I am not sure if this is even fixable, but it's actually not even possible to get the thermostat to 78F through google because of this.  If you say set it to 77F, it actually sets to 77F.   Minor glitch, but readings can/will be off.  If current temperature in the home is 78F (per thermostat), google will say it's 79F.  This is just examples, it would mess up on a lot of numbers!

**Comment by oznu** [comment]
@meluvalli 

I've removed all rounding work that the homebridge-gsh was doing to temperature values, beware that HomeKit still does rounding and depending on the plugin this may not fix this issue for you. For example, I'm not able to set my thermostat to 78°F in the **iOS Home app** either when the `minStep` value is set to 1 in the plugin. The option just isn't there, however when I setup a dummy thermostat that had `minStep` set to `0.1` I was able to choose 78°F in both the Google Home app and the iOS Home app and have the values sync both ways correctly.

Update v1.1.1 is now out, and a new option has been added to force °F.

![image](https://user-images.githubusercontent.com/3979615/71081208-dfd7aa80-21e2-11ea-80b7-f29a905be6e8.png)
